music_assistant/providers/listenbrainz_scrobble/__init__.py

```diff
@@ -0,0 +1,173 @@
+"""Allows scrobbling of tracks with the help of liblistenbrainz."""
+
+# icon.svg from https://github.com/metabrainz/design-system/tree/master/brand/logos
+# released under the Creative Commons Attribution-ShareAlike(BY-SA) 4.0 license.
+# https://creativecommons.org/licenses/by-sa/4.0/
+
+import asyncio
+import time
+from collections.abc import Callable
+from typing import Any
+
+from liblistenbrainz import Listen, ListenBrainz
+from music_assistant_models.config_entries import (
+    ConfigEntry,
+    ConfigValueType,
+    ProviderConfig,
+)
+from music_assistant_models.constants import SECURE_STRING_SUBSTITUTE
+from music_assistant_models.enums import ConfigEntryType, EventType
+from music_assistant_models.errors import SetupFailedError
+from music_assistant_models.event import MassEvent
+from music_assistant_models.playback_progress_report import MediaItemPlaybackProgressReport
+from music_assistant_models.provider import ProviderManifest
+
+from music_assistant.mass import MusicAssistant
+from music_assistant.models import ProviderInstanceType
+from music_assistant.models.plugin import PluginProvider
+
+
+async def setup(
+    mass: MusicAssistant, manifest: ProviderManifest, config: ProviderConfig
+) -> ProviderInstanceType:
+    """Initialize provider(instance) with given configuration."""
+    return ListenBrainzScrobbleProvider(mass, manifest, config)
+
+
+class ListenBrainzScrobbleProvider(PluginProvider):
+    """Plugin provider to support scrobbling of tracks."""
+
+    _client: ListenBrainz = None
+    _currently_playing: str | None = None
+    _on_unload: list[Callable[[], None]] = []
+
+    def _get_config(self) -> dict[str, ConfigValueType]:
+        return {
+            CONF_AUTH_TOKEN: self.config.get_value(CONF_AUTH_TOKEN),
+        }
+
+    async def loaded_in_mass(self) -> None:
+        """Call after the provider has been loaded."""
+        await super().loaded_in_mass()
+
+        if not self.config.get_value(CONF_AUTH_TOKEN):
+            self.logger.info("No auth token available, don't forget to configure!")
+            return
+
+        self._client = _get_client(self._get_config())
```

Comment:
Is there a reason for doing the setup here instead of in setup ? This way the provider will never fail (and auto retry) on login error or server temporary outages etc.

Reply:
I suggest to do the initialization in the async_setup method

Reply:
Inherited from last.fm scrobbler, where I guess it is part of the more complicated auth flow's requirements. Will simplify as you suggest.

Reply:
Did you mean `async def setup(self, config: CoreConfig) -> None:` rather than `async_setup`?

Reply:
`async def setup` seems to work so I assume that's what you meant, change pushed.